awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -0,0 +1,372 @@
+        except(Exception) as e:
+            self.fail_json(msg="Failed to read response body: {0}".format(e))
+
+        response_json = {}
+        if response_body and response_body != '':
+            try:
+                response_json = loads(response_body)
+            except(Exception) as e:
+                self.fail_json(msg="Failed to parse the response json: {0}".format(e))
+
+        return {'status_code': response.status, 'json': response_json}
+
+    def authenticate(self, **kwargs):
+        if self.username and self.password:
+            # Attempt to get a token from /api/v2/tokens/ by giving it our username/password combo
+            # If we have a username and password we need to get a session cookie
+            login_data = {
+                "description": "Ansible Tower Module Token",
+                "application": None,
+                "scope": "write",
+            }
+            # Post to the tokens endpoint with baisc auth to try and get a token
+            api_token_url = (self.url._replace(path='/api/v2/tokens/')).geturl()
+
+            try:
+                response = self.session.open(
+                    'POST', api_token_url,
+                    validate_certs=self.verify_ssl, follow_redirects=True,
+                    force_basic_auth=True, url_username=self.username, url_password=self.password,
+                    data=dumps(login_data), headers={'Content-Type': 'application/json'}
+                )
+            except(Exception) as e:
+                # Sanity check: Did the server send back some kind of internal error?
+                self.fail_json(msg='Failed to get token: {0}'.format(e))
+
+            try:
+                response_json = loads(response.read())
+                self.oauth_token_id = response_json['id']
+                self.oauth_token = response_json['token']
+            except(Exception) as e:
+                self.fail_json(msg="Failed to extract token information from response: {0}".format(e))
+
+        # If we have neiter of these then we can try un-authenticated access
+        self.authenticated = True
+
+    def default_check_mode(self):
+        '''Execute check mode logic for Ansible Tower modules'''
+        if self.check_mode:
+            try:
+                result = self.get_endpoint('ping')
+                self.exit_json(**{'changed': True, 'tower_version': '{0}'.format(result['json']['version'])})
+            except(Exception) as excinfo:
+                self.fail_json(changed=False, msg='Failed check mode: {0}'.format(excinfo))
+
+    def update_if_needed(self, existing_item, new_item, handle_response=True, **existing_return):
+        for field in new_item:
+            existing_field = existing_item.get(field, None)
+            new_field = new_item.get(field, None)
+            # If the two items don't match and we are not comparing '' to None
+            if existing_field != new_field and not (existing_field in (None, '') and new_field == ''):
```

Comment:
Let me backref where the same things are done in tower-cli

https://github.com/ansible/tower-cli/blob/b3d49ce2cf4ff175da4fb5ed1bdef110a7c81083/tower_cli/models/base.py#L401

We need to think of what this loses compared to the old logic. One thing is the null-like fields. I suspect we may have a problem with the `limit` field specifically, where it needs to be set to empty string. We should be able to skip null values, but I don't think we can for empty string values.